Design/Fit functions in python.md

```diff
@@ -0,0 +1,179 @@
+# Working with fit functions in python
+
+Although fitting functions have some of the functionality exposed to python the use of it is very limited and functions are constructed via strings. Values of the optimised parameters get extracted from `TableWorkspace`s output from the Fit algorithm. Both function construction and extraction of the results are cumbersome and error prone. This document describes a design of a python fitting API aiming to overcome these problems. The main idea of the solution proposed here is to create and manipulate a C++ [`IFunction`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/IFunction.h) object from python directly via a thin wrapper python class `FitFunctionWrapper`.
+
+## Function construction
+
+For each concrete function type (Gaussian, Lorentzian, etc) there should exist a corresponding python function with the same name automatically generated in a way similar to the algorithm functions. A function object is then constructed by calling such a function (constructor) and is an instance of `FitFunctionWrapper`. For example
+```
+  g = Gaussian()
+```
+A call without passing in any initialisation arguments should create a function with default parameters and attributes.
+Initial values of parameters and attributes can be given via keyword arguments, for example
+```
+  p = Polynomial(attributes={'n': 3}, parameters={'A0': 1, 'A1': 2, 'A3': 3, 'A4': 4})
+```
+Parameter and attribute names can also be used as the keys:
+```
+  g = Gaussian(Height=1, Sigma=0.1)
+  
+  p = Polynomial(n=3, A0=1, A1=2, A3=3, A4=4)
+```
+The implementation must check each keyword argument if it's a parameter or an attribute by calling `IFunction::hasAttribute(name)`. All passed attributes must be set before setting parameters. There is need to be a provision for setting attributes in a particular order in case it is essential for a fitting function. This can be done for example by passing a list of name/value pairs (tuples or lists) to the `attributes` keyword argument.
```

Comment:
`CubicSpline` can be changed by combining all `X`s into a single vector attribute. Then there is no need for that order. But here is a more evil example: `CrystalFieldMultiSpectrum`. It has attributes that through which you define the types of the member functions. When you set those attributes the parameters and attributes of the new member become parameters and attributes of this instance of `CrystalFieldMultiSpectrum`.

At the moment I cannot see a simple way of doing this internally and in a generic way. What I am suggesting is just a simple mechanism that can be used when needed.

On the other hand you can pass only the independent attributes to the constructor and set the others later in any order you like.